### DISCUSSION TITLE
Shapes overlaps challenge

### DISCUSSION THREAD

**Discussion by matthieuharbich** [abey79/vsketch#277] [Show and tell]
Hey there!

First, thank you so much @abey79 for the library. It makes my plotting experience much more fun. 

My challenge is to draw a path, but not drawing it if the path go into a "shape zone". As filling shapes doesn't make much sense when plotting, what would be your approach for doing it ? 

To illustrate what I mean: 
Today, if I draw circle 1 and then 2 with vsk.circle(), I get:
<img width="482" alt="image" src="https://user-images.githubusercontent.com/46715709/175818459-d9e4f72e-e01a-499c-a16b-af3e5d7e250a.png">

However, what I would like to achieve is: 
<img width="494" alt="image" src="https://user-images.githubusercontent.com/46715709/175818478-5c2a8441-515e-4d45-8ea0-0360555f4e3d.png">

So that the path of circle 1 "stops" if it enters in circle 2. 

What I was thinking, but seems complicated, is to draw circle 2 with sin() cos() and bezier curves and store all inner coordinates in a matrice. Then, when I draw the second shape (circle 1 here), I check if it enters the first shape ( circle 2 here ) by looking into the matrice.

I would hope for a much easier way, if anyone of you know about how to solve that more easily than my approach ?

Thank you so much for your support,

Cheers,

Matthieu

**Comment by abey79** [comment]
Hey Matthieu,

This is known as _hidden line removal_. Unfortunately, vsketch doesn't readily support this, but there are various ways to achieve the same result.

## Using Shapely

Shapely is a great 2D geometry library for python, which vsketch uses internally for other things. It supports boolean operations, so they can be used to "cut" parts of some geometries to emulate hidden line removal.

```python
from shapely.geometry import Point

# create two circular polygons
circle1 = Point(0, 0).buffer(4)
circle2 = Point(3, 0).buffer(4)

# draw the boundary of circle1 after being subtracted by circle2
vsk.geometry(circle1.exterior.difference(circle2))

# draw the boundary of circle2
vsk.geometry(circle2.exterior)
```

## Using `occult`

[`occult`](https://github.com/LoicGoulefert/occult) is a [vpype](https://github.com/abey79/vpype) plug-in by @LoicGoulefert which implements hidden line removal. vsketch uses vpype internally and makes it available via the `vsk.vpype()` API.

```python
vsk.circle(0, 0, radius=4)
vsk.circle(3, 0, radius=4)

# this requires vpype-occult to be installed and should be called at the end of the draw() function
vsk.vpype(f"occult")
```

Installing `occult` can be done using the following command (if you used pipx to install vsketch):
```
pipx inject vpype-occult vsketch
```

Or using pip if you manage your virtual env manually:
```
pip install vpype-occult
```

## Using shapes

[Shapes](https://vsketch.readthedocs.io/en/latest/autoapi/vsketch/index.html#vsketch.Shape) are a newer, badly-documented feature of vsketch which enable boolean operations. They can be used to reproduce the desired design, but the common line will be drawn twice.

```python
# create a shape of a circle "eaten" by another, offset circle
shp = vsk.createShape()
shp.circle(0, 0, radius=4)
shp.circle(3, 0, radius=4, op="difference")

# draw that shape
vsk.shape(shp)

# draw a second circle
vsk.circle(3, 0, radius=4)
```

The issue if this approach can be highlighted by changing the last line to `vsk.circle(3.2, 0, radius=4)`, which yields this result:

<img width="458" alt="image" src="https://user-images.githubusercontent.com/49431240/175821242-b97ab104-190f-497f-bc7f-022a4e780c52.png">


I hope this answer your question.

PS: fellow Lausannois??? 🇨🇭

**Reply by matthieuharbich** [reply]
Hey!

Thank you so much for the well documented answers, and the promptness of your answer. 

I'll play with the different solutions you gave (especially 1 and 2). 

Again thank you, it's really appreciated and perfectly answers my question. 

Cheers!

( Indeed I'm Lausannois! And don't want to be creepy but I think I briefly met you once in your house "sous gare", I started a PhD with Martin... ;) )

**Reply by abey79** [reply]
Hilarious :rofl:

**Reply by abey79** [reply]
I just added an [example](https://github.com/abey79/vsketch/tree/master/examples/hidden_line_removal) covering the three techniques.
